x-pack/plugin/eql/src/main/antlr/EqlBase.g4

```diff
@@ -200,6 +200,7 @@ STRING
     | '"'   ('\\' [btnfr"'\\] | ~[\r\n"\\])* '"'
     | '?"'  ('\\"' |~["\r\n])* '"'
     | '?\'' ('\\\'' |~['\r\n])* '\''
+    | '"""' ('\\"' |~[\r\n])*? '"""'
```

Comment:
I think this could really come to bite us. Mostly because it means that a raw string is not a raw string. With this, there's no way to express a literal `\"` anywhere in the string.

This branch allows a literal `\"`:
https://github.com/matriv/elasticsearch/compare/replace-unescaped-triple-doublequotes...rw-access:raw-string-changes

Reply:
This grammar has other issues:
```
"""""hello\"""world!\"""" == """"foo"\""bar""\"""" => ""hello"""world!" = "foo"""bar"""
```
=>
```
org.elasticsearch.xpack.eql.parser.ParsingException: line 1:21: token recognition error at: '!\'
```

and with the proposed solution you can have a `\"` literal if you double escape it `\\"`

Reply:
We could replace `\"` with `"` only if part of an escaped triple double quote sequence, but I think that produces more confusion for a user.

Reply:
but it _should_ give that syntax error because it's invalid syntax. the string was completed after seeing `"""`.

that sounds like the correct behavior. If we introduce any escapes at all we defeat the purpose of this being a raw string, and require some potentially complex parsing logic and strange edge cases

Reply:
`"""""hello\"""world!\""""` it's a valid syntax, it has a closing `"""`.
The previous `?"` couldn't accept any `"` and every `"` should be escaped.
With this approach we only need to escape a `"""` which is very rare to occur.

Reply:
but at the same time... Doesn't this make it impossible to end a string with a backslash? Or to end with ab unescaped quote?

doesn't seem worth the trade-off

Reply:
the thing I'm the most concerned with -- escape sequences mean that you could paste a string in here, and there's a decent chance that it is accepted but matches something entirely different, if it contains a `\"` anywhere in the original text

Reply:
I wonder if this is also an edge case:

how do these compare? I think with the current pattern with the `\"` escape, the first one is valid syntax but the second one is not?
```
"""my raw string\""" == "abc"
"""my raw string\""" == """abc"""
```

Reply:
![Screenshot_20201002-051836~2.png](https://user-images.githubusercontent.com/31489089/94907922-22126580-045e-11eb-9d69-ce6c45c87a0b.png)

Reply:
I can't think of a better solution myself. The `"` is something you need to escape even with the current syntax of `?"`.
The current proposal with `"""` allows to have unescaped `"` within the string apart from the last char where it needs to be escaped. Using another character to escape `"` like double `""` won't work and to my eyes seems more weird for a raw string. I think I have exhaustive examples now in the tests, so I'm happy if you want to also try with different patterns in the grammar if we can end up with some better behaviour.

Reply:
There are a number of issues at hand:
1. `"` is a common character. To avoid escaping it too often, the raw string relies on 3 triple quotes `"""`. This however can become a problem at the end of the raw string where a trailing `"` can mess up quoting
2. `"` are allowed inside and can be escaped to avoid closing a string `"""` through `\`. This again is it occurs at the end of the string.

To wit:
1. `""" dir "c:\dir""""` <-- 4 ending quotes
2. `""" dir c:\dir\"""`  <-- \ escapes the `"""`

Due to file paths on windows 2, aka `\"` is likely a common occurrence.
I see several options to solve this:

A. Require raw strings to not end with `"` and advice folks to add a space for example:
`""" dir "c:\dir" """`
B. Use a different escape character from `\`, say `|`. It solves 2 but not 1.
C. Do *not* allow escaping inside raw strings. This solves 2 but not 1.
D. Use a different set for quoting characters. Say "?""` and `""?`. However this does not solve the problem, rather just minimizes the occurrences in which they appear. Which might be a win in itself.

Considering `\` is traditionally used for escaping and in Window paths, I have a preference for C over B. That's because introducing a new character for escaping is surprising and might backfire again depending on the raw string .
B however mean that one cannot escape `"""` inside a raw string but then for that there's the regular string definition. Unfortunately this doesn't solve 1 either...

Reply:
A. I thing introducing a mandatory whitespace can lead to other issues, and is it only one space? what happens with more spaces, we keep all but one?, or with tabs?
B. I also don't like a new escape char, doesn't address the main issue no1.
D. I definitely don't like `D` since it still contains the `?` char which usually refers to regex or query param.

C. I'm ok with C so if a user needs to escape chars or needs a `"""`, he/she has to resort to the normal `"....."` syntax and properly escape what's needed.

Reply:
option E:

Allow for two optional additional trailing quotes with no escape sequences. then you can solve the trailing quotes issue without escapes.

```
"""[\r\n]*?""""?"?
```
![Screenshot_20201002-055915~2.png](https://user-images.githubusercontent.com/31489089/94911668-b8955580-0463-11eb-8c1f-14a5275a7e51.png)


and it doesn't suffer from these contextual issues with `\"`
![Screenshot_20201002-055234~2.png](https://user-images.githubusercontent.com/31489089/94911393-4f154700-0463-11eb-9f15-7d13e957270b.png)

Reply:
confirmed that `"""` isn't allowed anywhere in the string but `""` is;

![Screenshot_20201002-060749~2.png](https://user-images.githubusercontent.com/31489089/94912478-f2b32700-0464-11eb-8d92-c5c8b469d388.png)

Reply:
@rw-access I'm also happy with that, pushed the change and fixed all relevant tests.